keps/sig-network/2104-reworking-kube-proxy-architecture/README.md

```diff
@@ -0,0 +1,1003 @@
+- include equivalent implementations of in-project ones (userland, iptables and ipvs)
+- create reusable library elements for some proxy logic, such as conntrack cleaning, which might be called from different proxy implementations
+
+### Non-Goals
+
+- provide equivalent implementations of all third-party proxiers
+- force new users to run kpng in two separate daemons - this has performance benefits and is optional
+- force all proxiers to use the fullstate model: this is ideal for new implementations (easier to understand, no magic caches) but its not necessary.
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Rewrite the kube-proxy to be a "localhost" gRPC API provider that will be accessible as usual via
+TCP (`127.0.0.1:12345`) and/or via a socket (`unix:///path/to/proxy.sock`).
+
+- it will connect to the API server and watch resources, like the current proxy;
+- then, it will process them, applying Kubernetes specific business logic like topology computation
+  relative to the local host;
+- finally, provide the result of this computation to client via a gRPC watchable API.
+
+This decoupling allows kube-proxy and implementation to evolve in their own timeframes. For
+instance, introducing optimizations like EndpointSlice or new business semantics like Topology
+does not trigger a rebuild/release of any proxy implementation.
+
+In this implementation, we either: 
+- send the *full state*  of the kubernetes networking state-space to a client, every time anything needs to change.  Since this is done over GRPC or in memory, the bandwidth costs are low (anecdotally this has been measured, and it works for 1000s of serviecs and pods - we can attach specific results to this KEP as needed).  An example of this can be seen in the ebpf and nft proxies in the KPNG project (https://github.com/kubernetes-sigs/kpng/blob/master/backends/nft/nft.go).
+```
+// the entire statestpace of the Kubernetes networking model is embedded in this struct
+func Callback(ch <-chan *client.ServiceEndpoints) {
+	svcCount := 0
+	epCount := 0
+```
+- This implementation leverages a "DiffStore", which allows arbitrary, generic go objects to be diffed in memory by a backend.  This can be viewed at https://github.com/kubernetes-sigs/kpng/tree/master/client/diffstore.    The overall usage of this store is relatively intuitive: Write to it continuously, and only register "differneces" when looking at the Diffs.  The "Reset()" function causes the second "wave" in a series of writes to take place, such that a subsequent call to see the diff at a later time will reveal differences between the first and second series of writes.  Note that the `Get` call here will write a key if empty.
```

Comment:
(typo "differneces")

This text should go elsewhere (eg, after the list, or in a new section somewhere). You preceded the bulleted list with "In this implementation, we either:", so the reader is expecting to see option B at this point, not more details on option A.

Reply:
fixed , pushingnow